fix #18149 - insert measure after selection (don't move keysig, timesig and clef)

sammik opened the pull request:
Resolves: #18149 <!-- Replace `NNNNN` with a GitHub issue number, or a direct link if the issue is not on GitHub -->

<!-- Add a short description of and motivation for the changes here -->
Insert after selection now behaves differently, comparing to insert before
It doesn't move clefs, key and time signatures

`InsertMeasureOptions` already have `moveSignaturesClef` property, but code for Clefs were broken, it is fixed now.

(~~It doesn't handle barlines, as they are inconsitent also in "insert measure before selection" - "standard" barlines behave differently, than repeat barlines.~~)

Also barlines are correctly moved now.

<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)


https://github.com/musescore/MuseScore/assets/1646034/786b297c-4570-4c37-b914-75e923abdff6

cbjeukendrup commented on src/engraving/dom/edit.cpp:
```diff
@@ -4069,83 +4069,107 @@ MeasureBase* Score::insertMeasure(ElementType type, MeasureBase* beforeMeasure,
                         if (ps && ps->enabled()) {
                             EngravingItem* pc = ps->element(staffIdx * VOICES);
-                            if (pc) {
+                            if (pc && !options.moveSignaturesClef) {
```
It looks like we can move the `!options.moveSignaturesClef` a few `if` conditions up, to save some function calls.

cbjeukendrup commented on src/engraving/dom/edit.cpp:
```diff
@@ -4069,83 +4069,107 @@ MeasureBase* Score::insertMeasure(ElementType type, MeasureBase* beforeMeasure,
-                                undoRemoveElement(s);
-                            }
+                    } else {
+                        if (pm->repeatEnd()) {
```
Here you are using `pm` outside the `if (pm)` condition, so it might be null.

cbjeukendrup reviewed (commented):
I had a couple of remarks, but to save you some time I fixed them myself. Could you please check if everything looks right? 

(It's the most convenient to review all my commits separately, and to use the "hide whitespace" setting.)

sammik commented:
@cbjeukendrup 
Thank you, wery good changes, I had just one think to change, I posted review.

But if we doing this optimisations, may be it would be better also to have just one `staves` loop instead of three.

I post one other commit with proposed changes, please have a look.

cbjeukendrup commented:
@sammik Actually I made the for-loops consciously as local as possible, so that those `if` conditions are evaluated only once instead of for every staff again... what do you think?

sammik commented:
whops, 

> @sammik Actually I made the for-loops consciously as local as possible, so that those `if` conditions are evaluated only once instead of for every staff again... what do you think?

whops, sorry
@cbjeukendrup You are right, of course!

so, finaly, I changed only that one thing, I rewieved above and added two nullchecks

thanks for patience, please check it now

cbjeukendrup commented:
(One more push, to rebase, and to fix a mistake I had made in the first refactor commit (that you had also fixed in your refactor commit, but I wanted to make sure that my commit was "self-contained"))

cbjeukendrup commented:
I found one potential inconsistency:
- when you select the measure before an "end repeat" barline in the middle of the score, "after selection" will place the new measure before the end repeat barline
- when the _last_ measure of the score has an "end repeat" barline, and you do "after selection" on that last measure, the new measure will be placed _after_ the end repeat barline.

sammik commented:
> I found one potential inconsistency:
> 
> * when you select the measure before an "end repeat" barline in the middle of the score, "after selection" will place the new measure before the end repeat barline
> * when the _last_ measure of the score has an "end repeat" barline, and you do "after selection" on that last measure, the new measure will be placed _after_ the end repeat barline.

Yes, thanks, Ill trake a look.

Should be "after last bar" same as "append to the and of the score"? (Both move repeat barline to the and of the score?)

sammik commented:
@cbjeukendrup done

cbjeukendrup commented:
@sammik Sorry, I found one more issue... namely, the same goes for non-repeat barlines. If an end barline is customised, "insert after" will put the new measure after the custom barline, and if any other barline is customised, the new measure is placed before it.

cbjeukendrup commented:
Another one: if you select the title frame, then use Add > Measures > After selection, the newly inserted bar is placed before the initial key and time signatures. 
<img width="371" alt="Scherm­afbeelding 2023-09-29 om 22 30 27" src="https://github.com/musescore/MuseScore/assets/48658420/8f548577-d6bc-4ec4-a7e0-d62db858b807">

cbjeukendrup commented:
> Should be "after last bar" same as "append to the and of the score"? (Both move repeat barline to the and of the score?)

I think what you did now is fine. This way, the user can achieve both options.

sammik commented:
@cbjeukendrup all done :)
